File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/Views/StatsView/StatsView.m
//
//  StatsView.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 QuickBlox. All rights reserved.
//

#import "StatsView.h"
#import "CallParticipant.h"
#import "CallInfo.h"
#import "ActionsMenuView.h"
#import "MenuAction.h"

static NSString * const kStatsReportPlaceholderText = @"Loading stats report...";

@interface StatsView()
//MARK: - IBOutlets
@property (weak, nonatomic) IBOutlet UILabel *statsLabel;
@property (weak, nonatomic) IBOutlet UIButton *backButton;
@property (weak, nonatomic) IBOutlet UIButton *participantButton;
//MARK: - Properties
@property (strong, nonatomic) CallParticipant * _Nullable selectedParticipant;
@end


@implementation StatsView
//MARK - Setup
- (void)setCallInfo:(CallInfo *)callInfo {
    _callInfo = callInfo;
    self.selectedParticipant = self.callInfo.interlocutors.firstObject;
    
    __weak __typeof(self)weakSelf = self;
    [self.callInfo setOnChangedBitrate:^(NSNumber * _Nonnull ID, NSString * _Nonnull statsString) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if (![strongSelf.selectedParticipant.id isEqual:ID]) {
            return;
        }
        [strongSelf updateStats:statsString];
    }];
}

- (void)setSelectedParticipant:(CallParticipant *)selectedParticipant {
    _selectedParticipant = selectedParticipant;
    NSString *title = [NSString stringWithFormat:@"%@ ⌵", selectedParticipant.fullName];
    [self.participantButton setTitle:title forState:UIControlStateNormal];
}

//MARK: - Actions
- (IBAction)didTapBack:(UIButton *)sender {
    [self removeFromSuperview];
}

- (IBAction)didTapParticipant:(UIButton *)sender {
    ActionsMenuView *actionsMenuView = [[NSBundle mainBundle] loadNibNamed:@"ActionsMenuView" owner:nil options:nil].firstObject;
    for (CallParticipant *participant in self.callInfo.interlocutors) {
        participant.isSelected = [participant.id isEqual:self.selectedParticipant.id];
        __weak __typeof(self)weakSelf = self;
        MenuAction *selectParticipantAction = [[MenuAction alloc] initWithTitle:participant.fullName isSelected:participant.isSelected action:UserActionSelectParticipant handler:^(UserAction action) {
            __typeof(weakSelf)strongSelf = weakSelf;
            if ([strongSelf.selectedParticipant.id isEqual:participant.id]) {
                return;
            }
            strongSelf.selectedParticipant = participant;
            [strongSelf updateStats:kStatsReportPlaceholderText];
        }];
        [actionsMenuView addAction:selectParticipantAction];
    }
    [self addSubview:actionsMenuView];
    actionsMenuView.translatesAutoresizingMaskIntoConstraints = NO;
    [actionsMenuView.topAnchor constraintEqualToAnchor:self.topAnchor].active = YES;
    [actionsMenuView.leftAnchor constraintEqualToAnchor:self.leftAnchor].active = YES;
    [actionsMenuView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor constant: -3.0f].active = YES;
    [actionsMenuView.rightAnchor constraintEqualToAnchor:self.rightAnchor].active = YES;
}

//MARK: - Private Methods
- (void)updateStats:(NSString *)stats {
    self.statsLabel.text = stats.length ? stats : kStatsReportPlaceholderText;
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Calls/Views/StatsView/StatsView.swift
//
//  StatsView.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 12/18/18.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit

struct StatsViewConstants {
    static let statsReportPlaceholderText = "Loading stats report..."
}


class StatsView: UIView {
    //MARK: - IBOutlets
    @IBOutlet weak var statsLabel: UILabel!
    @IBOutlet weak var backButton: UIButton!
    @IBOutlet weak var participantButton: UIButton!

    //MARK: - Properties
    var callInfo: CallInfo! {
        didSet {
            selectedParticipant = callInfo.interlocutors.first
            
            callInfo.onChangedBitrate = { [weak self] (participant) in
                if self?.selectedParticipant?.userID != participant.iD { return }
                self?.updateStats(participant.statsString)
            }
        }
    }

    private var selectedParticipant: CallParticipant? {
         didSet {
             participantButton.setTitle("\(selectedParticipant?.fullName ?? "User") ⌵", for: .normal)
         }
     }

    //MARK: - Actions
    @IBAction func didTapBack(_ sender: UIButton) {
        removeFromSuperview()
    }

    @IBAction func didTapRarticipant(_ sender: UIButton) {
        let actionsMenuView = ActionsMenuView.loadNib()
        for participant in callInfo.interlocutors {
            participant.isSelected = participant.userID == self.selectedParticipant?.userID
            let selectParticipantAction = ActionMenu(title: participant.fullName,
                                                     isSelected: participant.isSelected,
                                                     action: .selectParticipant) { [weak self] action in
                if self?.selectedParticipant?.userID == participant.userID {
                    return
                }
                self?.selectedParticipant = participant
                self?.updateStats(StatsViewConstants.statsReportPlaceholderText)
            }
            actionsMenuView.addAction(selectParticipantAction)
        }
        addSubview(actionsMenuView)
        actionsMenuView.translatesAutoresizingMaskIntoConstraints = false
        actionsMenuView.rightAnchor.constraint(equalTo: rightAnchor, constant: 0.0).isActive = true
        actionsMenuView.topAnchor.constraint(equalTo: topAnchor, constant: 0.0).isActive = true
        actionsMenuView.leftAnchor.constraint(equalTo: leftAnchor, constant: 0.0).isActive = true
        actionsMenuView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: 0.0).isActive = true
    }
    
    //MARK: - Private Methods
    private func updateStats(_ stats: String?) {
        statsLabel.text = stats ?? StatsViewConstants.statsReportPlaceholderText
    }
}
